pretyping/canonicalStructures.mli

```diff
@@ -0,0 +1,156 @@
+(************************************************************************)
+
+open Constr
+
+(** Operations concerning records and canonical structures *)
+
+(** {6 Records } *)
+(** A structure S is a non recursive inductive type with a single
+   constructor (the name of which defaults to Build_S) *)
+
+module Projection : sig
+type t = {
+  name : Names.Name.t;
+  true_projection : bool;
+  canonical : bool;
+  body : Names.Constant.t option;
+}
+
+end
+
+module Structure : sig
+
+type t = {
+  record : Names.inductive;
+  projections : Projection.t list;
+  nparams : int;
+}
+
+val build : Environ.env -> Names.inductive -> Projection.t list -> t
+
+end
+
+val register_structure : Structure.t -> unit
+val subst_structure : Mod_subst.substitution -> Structure.t -> Structure.t
+val rebuild_structure : Environ.env -> Structure.t -> Structure.t
+
+(** [lookup_structure isp] returns the Structure.t associated to the
+   inductive path [isp] if it corresponds to a structure, otherwise
+   it fails with [Not_found] *)
+val lookup_structure : Names.inductive -> Structure.t
+
+(** [lookup_projections isp] returns the projections associated to the
+   inductive path [isp] if it corresponds to a structure, otherwise
+   it fails with [Not_found] *)
+val lookup_projections : Names.inductive -> Names.Constant.t option list
+
+(** raise [Not_found] if not a projection *)
+val find_projection_nparams : Names.GlobRef.t -> int
+
+(** raise [Not_found] if not a structure projection *)
+val find_projection : Names.GlobRef.t -> Structure.t
+
+val is_projection : Names.Constant.t -> bool
+
+(** Sets up the mapping from constants to primitive projections *)
+val register_primitive_projection : Names.Projection.Repr.t -> Names.Constant.t -> unit
+
+val is_primitive_projection : Names.Constant.t -> bool
+
+val find_primitive_projection : Names.Constant.t -> Names.Projection.Repr.t option
+
+(** {6 Canonical structures } *********************************************** *)
+
+(** A canonical structure declares "canonical" conversion hints between
+    the effective components of a structure and the projections of the
+    structure *)
+
+(** A ValuePattern.t characterizes the form of a component of canonical structure *)
+
+module ValuePattern : sig
+type t =
+    Const_cs of Names.GlobRef.t
+  | Proj_cs of Names.Projection.Repr.t
+  | Prod_cs
+  | Sort_cs of Sorts.family
+  | Default_cs
+
+val equal : Environ.env -> t -> t -> bool
+val print : t -> Pp.t
+
+(** Return the form of the component of a canonical structure *)
+val of_constr : Environ.env -> constr -> t * int option * constr list
+
+end
+
+type cs
+
+val check_and_decompose_canonical_structure :
+  Environ.env -> Evd.evar_map -> Names.GlobRef.t -> cs
+
+val register_canonical_structure :
+  warn:bool -> Environ.env -> Evd.evar_map -> cs -> unit
+
+val subst_canonical_structure : Mod_subst.substitution -> cs -> cs
+val solution_of_canonical_structure : cs -> Names.GlobRef.t
+
```

Comment:
the _canonical_structure maybe is a bit redundant if you would like to remove it.